lint/src/main/java/permissions/dispatcher/NoCorrespondingNeedsPermissionDetector.java

```diff
@@ -0,0 +1,113 @@
+package permissions.dispatcher;
+
+import com.android.tools.lint.detector.api.Category;
+import com.android.tools.lint.detector.api.Detector;
+import com.android.tools.lint.detector.api.Implementation;
+import com.android.tools.lint.detector.api.Issue;
+import com.android.tools.lint.detector.api.JavaContext;
+import com.android.tools.lint.detector.api.Scope;
+import com.android.tools.lint.detector.api.Severity;
+
+import java.util.EnumSet;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+import lombok.ast.Annotation;
+import lombok.ast.AstVisitor;
+import lombok.ast.CompilationUnit;
+import lombok.ast.ForwardingAstVisitor;
+import lombok.ast.Node;
+
+public class NoCorrespondingNeedsPermissionDetector extends Detector implements Detector.JavaScanner {
+
+    public static final Issue ISSUE = Issue.create("NoCorrespondingNeedsPermission",
+            "No corresponding @NeedsPermission found",
+            "The processor would ignore @OnShowRationale annotation if there is no corresponding @NeedsPermissions. For instance, if there is @OnShowRationale(Manifest.permission.CAMERA), then there must be @NeedsPermission(Manifest.permission.CAMERA) in the same class.",
+            Category.CORRECTNESS,
+            4,
+            Severity.ERROR,
+            new Implementation(NoCorrespondingNeedsPermissionDetector.class, EnumSet.of(Scope.JAVA_FILE)));
+
+    @Override
+    public AstVisitor createJavaVisitor(JavaContext context) {
+        return new AnnotationChecker(context);
+    }
+
+    static class AnnotationChecker extends ForwardingAstVisitor {
+        private final JavaContext context;
+        private Set<String> needsPermissionName;
```

Comment:
I guess `needsPermissionName` and `onShowRationaleName` can be initialized like below.

```java
private static final Set<String> NEEDS_PERMISSION_NAME = new HashSet<>() {{;
  add("OnShowRationale");
  add("permissions.dispatcher.OnShowRationale");
}};


```